File: 0140-word-break-ii/0140-word-break-ii.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        # Convert wordDict to a set for O(1) lookups
        word_set, self.n = set(wordDict), len(s)
        ans = []
        
        # Start the backtracking process
        self._backtrack(s, word_set, [], ans, 0)

        return ans
    
    def _backtrack(self, s: str, word_set: set, curr: List[str], ans: List[str], start: int) -> None:
        # If we've reached the end of the string, add the current sentence to results
        if start == self.n:
            ans.append(" ".join(curr))
            return
        
        # Iterate over possible end indices
        for end in range(start + 1, self.n + 1):
            word = s[start:end]
            
            # If the word is in the set, proceed with backtracking
            if word in word_set:
                curr.append(word)
                # Recursively call backtrack with the new end index
                self._backtrack(s, word_set, curr, ans, end)
                
                # Remove the last word to backtrack
                curr.pop()
                

File: 0140-word-break-ii/0140-word-break-ii.cpp
class Solution {
private:
    int n;
    void backtrack(const string& s, const unordered_set<string>& wordSet, string& curr, vector<string>& ans, int start){
        // If we've reached the end of the string, add the current sentence to ans
        if (start == n){
            ans.push_back(curr);
            return;
        }
        
        // Iterate over possible end indices
        for (int end = start + 1; end <= n; ++end){
            string word = s.substr(start, end - start);
            
            // If the word is in the set, proceed with backtracking
            if (wordSet.find(word) != wordSet.end()){
                string original = curr;
                if (!curr.empty()) curr += " ";
                curr += word;
                // Recursively call backtrack with the new end index
                backtrack(s, wordSet, curr, ans, end);
                curr = original;
            }
        }
    }
    
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        // Convert wordDict to a set for O(1) lookups
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        this -> n = s.size();
        vector<string> ans;
        string curr;
        
        // Start the backtracking process
        backtrack(s, wordSet, curr, ans, 0);
        
        return ans;
    }
};